### ISSUE TITLE
False-positive protobuf-type-error on scalar values inferred as instances of builtin types

### ISSUE THREAD

**Issue by zoedberg** [nelfin/pylint-protobuf#40]
Using pylint-protobuf `v0.18.4` I receive some new lint errors (with respect to `v0.17.0`) which I don't think are correct.

e.g.
`project/file.py:631: [E5903(protobuf-type-error), my_method] Field "MyRequest.my_request_field" is of type 'int' and value 'int()' will raise TypeError at runtime`

here IMO there's nothing wrong, the field should be an `int` and the code gives an `int`

`project/filepy:189: [E5903(protobuf-type-error), my_method] Field "MyResponse.my_response_field" is of type 'float' and value 'int()' will raise TypeError at runtime`

also here IMO there's nothing bad in providing an `int` parameter in a `float` field, I'm pretty sure the conversion from the two types is already handled by protobuf internally

`project/lfile.py:253: [E5903(protobuf-type-error), my_method] Field "MyRequest.my_request_field" is of type 'str' and value None will raise TypeError at runtime`

every protobuf parameter, of whatever type (also `bool`), can be set to `None`. The value will be considered not set and therefore it will be set to the default value (e.g. `False` for booleans, empty string for strings, etc.)

**Comment by nelfin** [comment]
Hi @zoedberg, thanks for the report. Type checking of fields was added in the 0.18.x release so these examples you've provided help work out the bugs. It looks like there might be a few different things in here so I'll try and separate them out.

Regarding the first error, do you have an example you could share? The only way I could replicate this message is via keyword arguments to a protobuf message constructor (see https://github.com/nelfin/pylint-protobuf/blob/master/pylint_protobuf/__init__.py#L186-L198 as the only places that raise messages of the format `x()`, all of the other `protobuf-type-error` messages reference the values directly), and if I pass the type "__builtins__.int" (not a *value* of type int) then the message is arguably correct (though I suppose its wording could be improved):

```protobuf
// example.proto
syntax = "proto3";
message Example {
  int32 value = 1;
  float fraction = 2;
}
```

```python
# issue40a.py
from example_pb2 import Example
e = Example(value=int)
```

```
$ pylint issue40a.py
************* Module issue40a
issue40a.py:2:4: E5903: Field "Example.value" is of type 'int' and value 'int()' will raise TypeError at runtime (protobuf-type-error)
$ python issue40a.py
Traceback (most recent call last):
  File "issue40a.py", line 2, in <module>
    e = Example(value=int)
TypeError: <class 'int'> has type type, but expected one of: int, long
```

Regarding the second error, it looks like it's caused by the same type of thing as above, but it is the case that `int -> float` should work and doesn't:

```python
# issue40b.py
from example_pb2 import Example
Example(fraction=123)
```

```
$ pylint issue40b.py
************* Module issue40b
issue40b.py:2:0: E5903: Field "Example.fraction" is of type 'float' and value 123 will raise TypeError at runtime (protobuf-type-error)
$ python issue40b.py  # no error
```

However, `float -> int` conversions do not work with protobuf (despite working with truncation in rregular python), so this needs some careful handling:

```python
# issue40c.py
from example_pb2 import Example
Example(value=123.0)
```

```
$ pylint issue40c.py
************* Module issue40c
issue40c.py:3:0: E5903: Field "Example.value" is of type 'int' and value 123.0 will raise TypeError at runtime (protobuf-type-error)
$ python issue40c.py
Traceback (most recent call last):
  File "issue40c.py", line 3, in <module>
    Example(value=123.0)
TypeError: 123.0 has type float, but expected one of: int, long
```


As for the third error, I just forgot about `optional`/default/not-set fields in my handling of this so I'll fix that.

In summary:
* Can you clarify your examples with the `int()` messages?
* `float` fields should accept int. This is a bug, I'll raise another issue for this.
* Fields should be able to take `None`/not-set values. I'll raise another issue for this.

**Comment by zoedberg** [comment]
> Regarding the first error, do you have an example you could share?

Using your `example.proto`, I've created the following example:

```python
# issue40a.py
import sys
from example_pb2 import Example
var_to_split = 'abcdfeg:1234'
str_field, int_str_field = var_to_split.split(':')
try:
    int_val = int(int_str_field)
except Exception:
    sys.exit(1)
e = Example(value=int_val)
```
this leads to the mentioned error:
```console
pylint --load-plugins=pylint_protobuf issue40a.py
************* Module issue40a
issue40a.py:9:4: E5903: Field "Example.value" is of type 'int' and value 'int()' will raise TypeError at runtime (protobuf-type-error)
```

I hope this will help.

Thanks for the great support!

**Comment by nelfin** [comment]
That is interesting! It looks like I'm failing to account for all the cases of inference there. It only happens with kwargs, and it only happens with values inferred with a known type but an "unknown" value (astroid isn't quiet smart enough to infer that `int_str_field = == '1234'`). I've reduced your example down a bit further:

```python
from example_pb2 import Example
var_to_split = 'abcdfeg:1234'
_, var = var_to_split.split(':')
e = Example(value=int(var))
# passes: Example(value=int('1234'))
# passes: e = Example(); e.value = int(var)
```

**Comment by nelfin** [comment]
Unfortunately, it's pretty late where I am at the moment and my brain is shutting down so I'm going to take a look at this one in the morning.

**Comment by zoedberg** [comment]
thanks! no rush :)

**Comment by nelfin** [comment]
The latest change causes that particular test to pass. See if this fixes the issues you are seeing of that type by installing the latest via `pip install git+https://github.com/nelfin/pylint-protobuf@1922e19dd1e07de54ca5b5c70595ac63053737c2`

**Comment by zoedberg** [comment]
Works great, thanks!

Just FYI, now a correct type error message is formatted like this:
`myproject/myfile.py:596: [E5903(protobuf-type-error), MyMessage] Field "MyMessageRequest.my_param" is of type 'int' and value <class 'float'> will raise TypeError at runtime`
not sure you wanted to use `<class 'float'>` , but not an issue for me, the message is clear

Closing this

**Comment by nelfin** [comment]
> not sure you wanted to use <class 'float'> , but not an issue for me
Not an issue for me either, I think I'd rather it say "...and value of type float will raise TypeError..." but I'll do that later.

These changes have been released as `v0.18.5`.
